### ISSUE TITLE
access_token matching_token_for incorrect behaviour

### ISSUE THREAD

**Issue by foodforarabbit** [doorkeeper-gem/doorkeeper#941]
lib/doorkeeper/models/access_token_mixin.rb ln 98 

We get the last authorized token fine but then we return it only if the last token's scope matches the requested scope and ALL the application's scopes. This only makes sense and only works if we request client_credentials tokens with all the application's scopes. 

I think we should pass in Scopes.new instead of application.try(:scopes).

**Comment by maclover7** [comment]
Hi @foodforarabbit! Before we move forward, since this is a security sensitive project, can you explain if/how not checking against all of the application's scopes could be used a potential vulnerability in an application using Doorkeeper?

**Comment by foodforarabbit** [comment]
I was just searching the repo and I've noticed that someone has already submitted a pull request for this issue: https://github.com/doorkeeper-gem/doorkeeper/pull/848

Anyway to expand on the "problem" I'm not sure about the security implications but checking against all the possible scopes for a "matching token" seems to be unexpected behaviour. 

Assume I have an application with scopes 'public user_read user_write admin_read admin_write' and I have an existing token with scopes 'public user_read', it will not "match" a new request for an identical token because the matcher compares the token's scopes 'public user_read' against ALL the application's scopes.

Another fix that might be useful is on line 86 last_authorized_token_for(...) should probably also include scopes in the where clause

```
def last_authorized_token_for(application_id, resource_owner_id, scopes)
        where(application_id: application_id,
              resource_owner_id: resource_owner_id,
              revoked_at: nil,
              scopes: scopes.to_s).
          send(order_method, created_at_desc).
          limit(1).
          to_a.
          first
      end
```

The original way, if alternately request for a token with two different sets of scopes, you will never be able to "reuse" an access token even if it matches and isn't expired yet.

**Comment by Exelord** [comment]
Any news here? :)

**Comment by nbulaj** [comment]
It can be connected with #957 and #907, so it requires some exploration

**Comment by nbulaj** [comment]
Fixed in #907
